kernel/src/hil/public_key_crypto/keys.rs

```diff
@@ -283,3 +283,112 @@ pub trait RsaPrivKeyMut: PubPrivKeyMut + RsaKeyMut {
     /// the output of this function.
     fn take_exponent(&self) -> Option<&'static mut [u8]>;
 }
+
+/// Client for selecting keys.
+pub trait SelectKeyClient {
+    /// Called when the number of keys available is known.
+    fn get_key_count_done(&self, count: usize);
+
+    /// Called when the specified key is active and ready to use for the next
+    /// cryptographic operation.
+    ///
+    /// ### `error`:
+    ///
+    /// - `Ok(())`: The key was selected successfully.
+    /// - `Err(())`: The key was selected set successfully.
+    ///   - `ErrorCode::INVAL`: The index was not valid.
+    ///   - `ErrorCode::FAIL`: The key could not be set.
+    fn select_key_done(&self, index: usize, error: Result<(), ErrorCode>);
+}
+
+/// Interface for selecting an active key among the number of available keys.
+///
+/// This interface allows for the implementer to maintain an opaque internal
+/// representation of keys. They may be stored in memory, flash, or in secure
+/// element (where the actual key may not be accessible). Users of this
+/// interface can select which key is active for cryptographic operations. There
+/// is no assumption for implementers of this interface that keys can be added
+/// or changed or that keys can be specified by their actual bytes in a slice.
+///
+/// Selecting a key is asynchronous as it may require communication over a bus
+/// or waiting for an interrupt.
+///
+/// Keys are specified by an index starting at zero and going to
+/// `get_key_count()-1`, without gaps. Selecting a key between `0` and
+/// `get_key_count()-1` must not fail with `ErrorCode::INVAL`.
+///
+/// The stored keys can be public or private keys.
+pub trait SelectKey<'a> {
+    /// Return the number of keys that the device can switch among.
+    ///
+    /// Each key must be identifiable by a consistent index.
+    ///
+    /// This operation is asynchronous and its completion is signaled by
+    /// `get_key_count_done()`.
+    ///
+    /// ## Return
+    ///
+    /// `Ok()` if getting the count has started. Otherwise:
+    /// - `Err(ErrorCode::FAIL)` if the key count could not be started and there
+    ///   will be no callback.
+    fn get_key_count(&self) -> Result<(), ErrorCode>;
+
+    /// Set the key identified by its index as the active key.
+    ///
+    /// Indices start at 0 and go to `get_key_count() - 1`.
+    ///
+    /// This operation is asynchronous and its completion is signaled by
+    /// `select_key_done()`.
+    ///
+    /// ## Return
+    ///
+    /// `Ok()` if the key select operation was accepted. Otherwise:
+    /// - `Err(ErrorCode::INVAL)` if the index is not valid.
+    /// - `Err(ErrorCode::ALREADY)` if the key is already selected.
```

Comment:
What's the purpose of making this an error? Seems like it could just return `Ok()` even if it was a no-op underneath?

It feels like the only thing exposing this as an error does is potentially reveal across users which key a prior user had used to do something?

Reply:
Because if it returns Ok it has to issue a callback, which puts the burden on the implementer rather than the caller.

Reply:
I guess it could just set the key anyway.